src/main/java/javaslang/collection/List.java

```diff
@@ -426,7 +426,15 @@ default int lastIndexOf(T element) {
     @Override
     default Tuple2<List<T>, List<T>> partition(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        return Tuple.of(filter(predicate), filter(predicate.negate()));
+        List<T> left = nil(), right = nil();
+        for (T t : this) {
+            if (predicate.test(t)) {
+                left = left.prepend(t);
+            } else {
+                right = right.prepend(t);
+            }
+        }
+        return Tuple.of(left.reverse(), right.reverse());
```

Comment:
@szarnekow of course the ArrayLists have also to be passes. That means we need 3 passes instead of 1. I updated the comments above accordingly.

I see a slight improvement using ArrayList compared to the first solution using javaslang's linked List and reverse() because no intermediate Cons objects are created while interating the original List.